### PR TITLE
#171 : Show edited/moved marker on messages

### PR DESCRIPTION
**Pull request by fizy069**
Haven't added the zulip icons yet, since I wasn't able to run the script to auto generate the icons.dart file.

Fixes: #171

### CODE REVIEW

**Comment by fizy069** [comment]
CI failed probably because I did not add the generated files to the commit. Will do so soon.

**Comment by gnprice** [comment]
Thanks for the contribution!

In addition to getting the existing tests passing, before we can review this PR in detail you'll need to write tests for it — see the note in the README file.

If you have questions as you try to do that, we're happy to discuss in `#mobile-team` on chat.zulip.org.

**Comment by gnprice** [comment]
Please also post screenshots. That helps people review the UI changes.

**Comment by gnprice** [comment]
I've edited the PR description to link it up with the issue, to make it easy to find when looking at the issue.

When the PR is ready for review, just post a comment here saying so and then I can take a look.

**Comment by fizy069** [comment]
#F171
Here is how it looks currently.

Also, the PR is ready for review, could you please give your inputs?

File: lib/api/model/model.dart [hunk 1]
```diff
@@ -664,6 +665,37 @@ enum MessageFlag {
   String toJson() => _$MessageFlagEnumMap[this]!;
 }
 
+@JsonSerializable(fieldRename: FieldRename.snake)
+class MessageEditHistory {
+
+  final String? prevContent;
```
Comment (gnprice) [inline lib/api/model/model.dart:671]: nit:
```suggestion
class MessageEditHistory {
  final String? prevContent;
```

File: lib/api/model/model.dart [hunk 2]
```diff
@@ -664,6 +665,37 @@ enum MessageFlag {
   String toJson() => _$MessageFlagEnumMap[this]!;
 }
 
+@JsonSerializable(fieldRename: FieldRename.snake)
+class MessageEditHistory {
```
Comment (gnprice) [inline lib/api/model/model.dart:669]: nit: Let's put this at the end of this file, after `DmMessage`. That way it doesn't get between reading `Message` and reading `StreamMessage`, or `DmMessage`.

(Most of the time, the edit history isn't part of what one needs to think about when thinking about a message.)

File: lib/api/model/model.dart [hunk 3]
```diff
@@ -664,6 +665,37 @@ enum MessageFlag {
...
+
+  final String? prevContent;
+  final String? prevRenderedContent;
+  final int? prevRenderedContentVersion;
+  final int? prevStream;
+  final String? prevTopic;
```
Comment (gnprice) [inline lib/api/model/model.dart:675]: The API docs say this field was renamed in Zulip Server 5.0. This codebase currently still supports back to Zulip Server 4.0 — see the "Editing API types" section in the README. So this should handle the old name too.

To avoid the rest of the code needing to look for both names, we should have the class use only the new name, but have the deserialization code look for the old name if it doesn't find the new name. See `User.isSystemBot` for an example of how to do that.

(Then that will also be a case that should definitely get a test.)

File: lib/api/model/model.dart [hunk 4]
```diff
@@ -664,6 +665,37 @@ enum MessageFlag {
...
+    required this.prevStream,
+    required this.prevTopic,
+    required this.stream,
+    required this.timestamp,
+    required this.topic,
+    required this.userId,
+  });
+
+  factory MessageEditHistory.fromJson(Map<String, dynamic> json) =>
+      _$MessageEditHistoryFromJson(json);
```
Comment (gnprice) [inline lib/api/model/model.dart:694]: nit: two-space indents
```suggestion
  factory MessageEditHistory.fromJson(Map<String, dynamic> json) =>
    _$MessageEditHistoryFromJson(json);
```

File: lib/api/model/model.dart [hunk 5]
```diff
@@ -801,4 +835,4 @@ class DmMessage extends Message {
 
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
-}
+}
```
Comment (gnprice) [inline lib/api/model/model.dart:838]: This final line is incomplete. See https://github.com/zulip/zulip-flutter/pull/534#discussion_r1515353889 for discussion.

File: lib/api/model/model.dart [hunk 6]
```diff
@@ -569,7 +569,7 @@ sealed class Message {
   String content;
   final String contentType;
 
-  // final List<MessageEditHistory> editHistory; // TODO handle
+  final List<MessageEditHistory>? editHistory; // TODO handle
```
Comment (gnprice) [inline lib/api/model/model.dart:572]: This TODO comment should be removed — it's calling for us to do what this PR is doing :slightly_smiling_face:

File: lib/widgets/message_list.dart [hunk 7]
```diff
@@ -892,6 +893,16 @@ class MessageWithPossibleSender extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     final message = item.message;
+    bool messageEdited = false, messageMoved = false;
+    if(message.editHistory != null) {
+      if((message.editHistory?[message.editHistory!.length-1].topic !=
+          message.editHistory?[message.editHistory!.length-1].prevTopic) ||
```
Comment (gnprice) [inline lib/widgets/message_list.dart:899]: This is pretty messy to read. The repeated subexpression here should be pulled out as a local variable.

Also see the `Iterable.last` getter :-)

File: lib/widgets/message_list.dart [hunk 8]
```diff
@@ -892,6 +893,16 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+          (message.editHistory?[message.editHistory!.length-1].stream !=
+          message.editHistory?[message.editHistory!.length-1].prevStream)){
+            messageMoved = true;}
+      if(message.editHistory?[message.editHistory!.length-1].prevContent!=null){
+        messageEdited = true;
```
Comment (gnprice) [inline lib/widgets/message_list.dart:904]: What does this do if the message is edited, and then it's moved? What's the desired behavior in that case?

File: lib/api/model/model.dart (hunk 6 above)
Comment (gnprice) [inline lib/api/model/model.dart:572]: This new field needs to be kept up to date in response to server events. From #171:
> The first phase of work on this issue will be to start storing and maintaining that information. This will involve adding a field to the `Message` class, and then updating that field as appropriate in response to events.

File: windows/flutter/CMakeLists.txt [hunk 9]
```diff
@@ -10,6 +10,11 @@ include(${EPHEMERAL_DIR}/generated_config.cmake)
 # https://github.com/flutter/flutter/issues/57146.
 set(WRAPPER_ROOT "${EPHEMERAL_DIR}/cpp_client_wrapper")
 
+# Set fallback configurations for older versions of the flutter tool.
+if (NOT DEFINED FLUTTER_TARGET_PLATFORM)
+  set(FLUTTER_TARGET_PLATFORM "windows-x64")
```
Comment (gnprice) [inline windows/flutter/CMakeLists.txt:15]: This change looks unrelated to the rest of the PR. As part of making clear and coherent commits:
https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html
it'll therefore need to be its own commit, with a commit message explaining why we want to make this change.

**Review by gnprice — COMMENTED** [review]
Thanks for the revision!

Some quick comments below on the code so far.

One thing we'll need before we can fully review this PR is for the changes to be organized into [clear and coherent commits](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html). See the relevant README section:
https://github.com/zulip/zulip-flutter#submitting-a-pull-request

From the screenshots, it looks like this UI doesn't yet match the design in at least one important way: pay close attention to where the edited/moved markers are when at rest, before swiping to show the word "Edited" or "Moved".

File: lib/widgets/message_list.dart (hunk 8 above)
Comment (fizy069) [reply lib/widgets/message_list.dart:904]: If a message has been edited and moved, or has been moved and then edited - I believe it should prioritize showing edited over moved in both cases. Please correct me if I'm wrong.

Comment (gnprice) [reply lib/widgets/message_list.dart:904]: Yeah, that's right.
